gprofiler/system_metrics.py

```diff
@@ -0,0 +1,106 @@
+import statistics
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from threading import Event, RLock, Thread
+from typing import List, Optional
+
+import psutil
+
+from gprofiler.exceptions import ThreadStopTimeoutError
+
+DEFAULT_POLLING_INTERVAL_SECONDS = 5
+STOP_TIMEOUT_SECONDS = 30
+
+
+@dataclass
+class Metrics:
+    # The average CPU usage between gProfiler cycles
+    cpu_avg: Optional[float]
+    # The average RAM usage between gProfiler cycles
+    mem_avg: Optional[float]
+
+
+class SystemMetricsMonitorBase(metaclass=ABCMeta):
+    @abstractmethod
+    def start(self):
+        pass
+
+    @abstractmethod
+    def stop(self):
+        pass
+
+    @abstractmethod
+    def _get_average_memory_utilization(self) -> Optional[float]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_cpu_utilization(self) -> Optional[float]:
+        """
+        Returns the CPU utilization percentage since the last time this method was called.
+        """
+        raise NotImplementedError
+
+    def get_metrics(self) -> Metrics:
+        return Metrics(self._get_cpu_utilization(), self._get_average_memory_utilization())
+
+
+class SystemMetricsMonitor(SystemMetricsMonitorBase):
+    def __init__(self, stop_event: Event, polling_rate_seconds: int = DEFAULT_POLLING_INTERVAL_SECONDS):
+        self._polling_rate_seconds = polling_rate_seconds
+        self._mem_percentages: List[float] = []
+        self._stop_event = stop_event
+        self._thread = None
+        self._lock = RLock()
+
+        self._get_cpu_utilization()  # Call this once to set the necessary data
+
+    def start(self):
+        assert self._thread is None, "SystemMetricsMonitor is already running"
+        self._stop_event.clear()
```

Comment:
You shouldn't call `clear` or `set` on this event - you're not controlling it, you're merely using it. And you won't need to - by the time `stop` is called, the event will be already set. You can add `assert self._stop_event.is_set()`?